src/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs

```diff
@@ -11,6 +11,452 @@ namespace System.Net.Sockets.Tests
+        {
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, false));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, true));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, FileAccess.ReadWrite));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, FileAccess.ReadWrite, false));
+        }
+
+        [Fact]
+        public void Ctor_NotConnected_ThrowsIOException()
+        {
+            Assert.Throws<IOException>(() => new NetworkStream(new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp)));
+        }
+
+        [Fact]
+        public async Task Ctor_NotStream_ThrowsIOException()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                await client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port));
+                Assert.Throws<IOException>(() => new NetworkStream(client));
+            }
+        }
+
+        [Fact]
+        public async Task Ctor_NonBlockingSocket_ThrowsIOException()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = listener.AcceptAsync();
+                await Task.WhenAll(acceptTask, client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
+                using (Socket server = await acceptTask)
+                {
+                    server.Blocking = false;
+                    Assert.Throws<IOException>(() => new NetworkStream(server));
+                }
+            }
+        }
+
+        [Fact]
+        public async Task Ctor_Socket_CanReadAndWrite_DoesntOwn()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = listener.AcceptAsync();
+                await Task.WhenAll(acceptTask, client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
```

Comment:
This is a bit of a nit, but I don't love this WhenAll pattern.

The problem is that if the client.ConnectAsync fails for any reason, the WhenAll will hang the test, because the acceptTask will never complete.

On the other hand, if you replace this with:
await connectTask;
await acceptTask;

then if the connect fails, the test fails.